code/datums/components/chameleon/action.dm

```diff
@@ -0,0 +1,374 @@
+#define CHAMELEON_MANUAL_COOLDOWN 2 MINUTES
+
+/// The action for the chameleon panel.
+/datum/action/chameleon_panel
+	name = "Chameleon Outfit Panel"
+	button_icon_state = "chameleon_outfit"
+	COOLDOWN_DECLARE(next_manual)
+
+/datum/action/chameleon_panel/Trigger()
+	if(!IsAvailable())
+		return
+	ui_interact(owner)
+
+/datum/action/chameleon_panel/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "ChameleonPanel")
+		ui.set_autoupdate(TRUE)
+		ui.open()
+
+/datum/action/chameleon_panel/ui_state(mob/user)
+	return GLOB.not_incapacitated_state
+
+/datum/action/chameleon_panel/ui_static_data(mob/user)
+	var/mob/living/l_user = isliving(user) ? user : null
+	. = list(
+		"outfits" = list(),
+		"presets" = list(),
+		"icons" = list(
+			"outfits" = list()
+		)
+	)
+	var/list/assets_to_send = list()
+	for(var/outfit_path in subtypesof(/datum/outfit/job))
+		var/datum/outfit/outfit = outfit_path
+		if(initial(outfit.can_be_admin_equipped))
+			.["outfits"] += list(list(
+				"name" = initial(outfit.name),
+				"type" = "[outfit]"
+			))
+			var/outfit_asset = get_outfit_icon(outfit_path)
+			if(outfit_asset)
+				.["icons"]["outfits"]["[outfit]"] = SSassets.transport.get_asset_url(outfit_asset)
+				assets_to_send += outfit_asset
```

Comment:
The reason this isn't part of the spritesheet like the disguises, is because for some reason, job outfit icons _do not properly insert into a spritesheet_, resulting in around half of the job icons being blank.

Reply:
What was your original code which did that thing?